src/mpl/src/gpu/mpl_gpu_ze.c

```diff
@@ -816,12 +812,34 @@ static int munmapFunction(int nfds, int *fds, void *ptr, size_t size)
     mapped pointers can be from a remote IPC handle
 */
 
-static inline void free_ipc_handle_cache(MPL_ze_ipc_handle_entry_t * cache_entry)
+static inline int free_ipc_handle_cache(MPL_ze_ipc_handle_entry_t * cache_entry)
 {
+    int mpl_err = MPL_SUCCESS;
+    int ret = ZE_RESULT_SUCCESS;
+
     if (cache_entry->mapped_ptr) {
-        munmapFunction(cache_entry->nfds, cache_entry->fds, cache_entry->mapped_ptr,
-                       cache_entry->mapped_size);
+        munmapFunction(cache_entry->nfds, cache_entry->mapped_ptr, cache_entry->mapped_size);
+        if (!cache_entry->handle_cached) {
+            /* free the ipc handle itself */
+            ret = zeMemPutIpcHandle(ze_context, cache_entry->ipc_handle.ipc_handles[0]);
+            ZE_ERR_CHECK(ret);
+        }
     }
+
+    if (cache_entry->handle_cached) {
+        ret = zeMemPutIpcHandle(ze_context, cache_entry->ipc_handle.ipc_handles[0]);
+        ZE_ERR_CHECK(ret);
+    }
```

Comment:
I'm a bit confused by these two checks, can you clarify?

Reply:
We have two paths to create a cache entry in ipc_cache_tracked. One is calling zeMemGetIpcHandle in MPL_gpu_ipc_handle_create, another is implicitly when mmap a dmabuf_fd (in GDRcopy path) where we also call zeMemGetIpcHandle in order to get the dmabuf_fd. In the first case, handle_cached is true, and it stores original IPC handle together with the shareable IPC handle data, while in the second case, only the original IPC handle is saved, but not converted to the shareable IPC handle. These two checks just checked for those two paths.

Reply:
Ah I see now, thanks for clarifying.